op-verifier/contracts/lib/DisputeGameLookup.sol

```diff
@@ -0,0 +1,366 @@
+     * @param minAge Minimum age required.
+     */
+    error GameNotFound(uint256 minAge);
+
+    /**
+     * @dev Emitted when the dispute game is not enabled.
+     */
+    error DisputeGameNotEnabled();
+
+    /**
+     * @notice Internal function to get the dispute game factory from the optimism portal.
+     * @param optimismPortal The Optimism portal output root contract.
+     * @return factory The dispute game factory.
+     */
+    function _disputeGameFactory(
+        IOptimismPortalOutputRoot optimismPortal
+    ) internal view returns (IDisputeGameFactory) {
+        try optimismPortal.disputeGameFactory() returns (
+            IDisputeGameFactory factory
+        ) {
+            if (address(factory) == address(0)) {
+                revert DisputeGameNotEnabled();
+            }
+
+            return factory;
+        } catch {
+            revert DisputeGameNotEnabled();
+        }
+    }
+
+    /**
+     * @notice Retrieves the dispute game at the specified index.
+     * @param optimismPortal The Optimism portal output root contract.
+     * @param index The index of the dispute game.
+     * @param minAge The minimum age required to challenge the game.
+     * @param maxAge The maximum age allowed to challenge the game.
+     * @return outputRoot The root claim of the dispute game.
+     * @return gameType The type of the dispute game.
+     * @return gameCreationTime The creation time of the dispute game.
+     * @return proxy The dispute game proxy.
+     */
+    function getDisputeGame(
+        IOptimismPortalOutputRoot optimismPortal,
+        uint256 index,
+        uint256 minAge,
+        uint256 maxAge
+    )
+        internal
+        view
+        returns (
+            bytes32 outputRoot,
+            GameType gameType,
+            uint64 gameCreationTime,
+            IDisputeGame proxy
+        )
+    {
+        IDisputeGameFactory disputeGameFactory = _disputeGameFactory(
+            optimismPortal
+        );
+
+        // Get dispute game at index
+        Timestamp gameCreationTimeRaw;
+        (gameType, gameCreationTimeRaw, proxy) = disputeGameFactory.gameAtIndex(
+            index
+        );
+        outputRoot = proxy.rootClaim().raw();
+
+        // Unwrap gameCreationTime to uint64
+        gameCreationTime = gameCreationTimeRaw.raw();
+
+        // Wait for challenger to challenge the dispute game
+        if (block.timestamp - gameCreationTime < minAge) {
+            revert GameTooEarly(
+                index,
+                block.timestamp - gameCreationTime,
+                minAge
+            );
+        }
+
+        // Reject dispute game that has been expired
+        if (maxAge > 0 && block.timestamp - gameCreationTime > maxAge) {
+            revert GameExpired(
+                index,
+                block.timestamp - gameCreationTime,
+                maxAge
+            );
+        }
+
+        // Revert if the game is challenged
```

Comment:
We should also reject games that have an open challenge, not just games that have already been successfully challenged. Otherwise an attacker can keep using an invalid state during the whole challenge process.

Reply:
That would make it delay a few days at least if we check for the `IN_PROGRESS` status. I haven't seen a function to directly check for an open challenge.

Reply:
Can be seen at https://sepolia.etherscan.io/address/0x99b0e2a750Cd985A3707a8C792BB0ADb28d148b4#readContract

That try to implement this one

![image](https://github.com/user-attachments/assets/d4c819a6-5426-483c-b087-4d90775f8ce3)

![image](https://github.com/user-attachments/assets/8b8e6798-e554-49fd-a016-9b51c7682d51)

It go back 4 days in OP sepolia testnet

Reply:
I'm a bit confused by this. Reading the OP code, it looks like the assumption is that every dispute game in fact corresponds to an active dispute. Are dispute games automatically created for each new block committed, or are they only created when someone challenges a block?

Reply:
Dispute games are created for every state root proposal and the successful resolution of a game is what updates the anchor state in the AnchorStateRegistry.

In principle anyone can create a game (and Optimism do have confirmed production test canaries trying to break things with invalid root claims) but the game has to resolve in favour of the proposed state for the anchor to be updated.

https://github.com/ethereum-optimism/optimism/blob/c2f290d479f7319776205c2640f140a8423cb987/packages/contracts-bedrock/src/dispute/AnchorStateRegistry.sol#L59

Reply:
Okay, so there's no state in which a proposed state root has simply not been challenged by anyone? Is there a way to see if anyone has actually begun the dispute process?

Reply:
It may be possible to detect that the length of `claimData` is greater than 1, which means someone has tried to challenge, but I'm not 100% sure.

Reply:
Implemented open challenge detection with these checks

```solidity
    function _isGameChallenging(
        IChallengingDisputeGame proxy
    ) internal view returns (bool) {
        // If the game is finalized, we ignore the challenging status
        if (proxy.status() == GameStatus.DEFENDER_WINS) return false;

        // If claimData length is greater than 1, the game is challenging
        try proxy.claimData(1) {} catch {
            return true;
        }

        // If root claim is countered by someone then the game is challenging
        (, address counteredBy, , , , , ) = proxy.claimData(0);
        if (counteredBy != address(0)) return true;

        return false;
    }
```

And fixed interpolation search revert on case that lower bound has moved beyond the max timestamp

Reply:
Looks good! Is interpolation search working fine now then? Some tests (and benchmarks) are all that's missing to merge this, I think.

Thank you again for all your hard work on this PR.

Reply:
Here’s the revised version with improved grammar:

Test cases have been implemented in my [dispute-game-lookup](https://github.com/Opti-domains/dispute-game-lookup/tree/main) repository. I am using Foundry's fuzz testing combined with mainnet forking to identify and fix edge cases, including integer underflow and rounding errors.

The test cases can be found at the following links:
* [DisputeGameLookup](https://github.com/Opti-domains/dispute-game-lookup/blob/main/test/DisputeGameLookup.t.sol)
* [L2OutputOracleLookup](https://github.com/Opti-domains/dispute-game-lookup/blob/main/test/L2OutputOracleLookup.t.sol)

The files in the [dispute-game-lookup](https://github.com/Opti-domains/dispute-game-lookup/tree/main) repo and this repo differ only in the import section, as shown in these diffs:
* [DisputeGameLookup](https://www.diffchecker.com/99DREhBY/)
* [L2OutputOracleLookup](https://www.diffchecker.com/66xKoWgd/)
* [OPOutputLookup](https://www.diffchecker.com/Tug5lgIh/)

Reply:
For benchmarking, it take about 3.3s for each record processed in parallel on chrome network profiling

![image](https://github.com/user-attachments/assets/8fb786fd-8960-4bcd-a600-7f859001cb80)

https://app.ens.domains/disputegame.eth (On Sepolia)